docs/agent-development.md

```diff
@@ -384,3 +387,55 @@ Here's a list of the config options across all agents, their types, default valu
+
+### Agent Configuration via Kibana
+
+Also known as "central configuration". Agents can query the APM Server for configuration updates; the server proxies and caches requests to Kibana.
+
+Agents should poll the APM Server for config periodically by sending an HTTP request to the `/config/v1/agents` endpoint. Agents must specify their service name, and optionally environment. The server will use these to filter the configuration down to the relevant service and environment. There are two methods for sending these parameters:
+
+1. Using the `GET` method, pass them as query parameters: `http://localhost:8200/config/v1/agents?service.name=opbeans&service.environment=production`
+2. Using the `POST` method, encode the parameters as a JSON object in the body, e.g. `{"service": {"name": "opbeans", "environment": "production"}}`
+
+The server will respond with a JSON object, where each key maps a config attribute to a string value. The string value should be interpreted the same as if it were passed in via an environment variable. Upon receiving these config changes, the agent will update its configuration dynamically, overriding any config previously specified. That is, config via Kibana takes highest precedence.
+
+To minimise the amount of work required by users, agents should aim to enable this feature by default. This excludes RUM, where there is a performance penalty.
+
+#### Caching
+
+As mentioned above, the server will cache config for each unique `service.name`, `service.environment` pair. The server will respond to config requests with two related response headers: [Etag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) and [Cache-Control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control).
+
+Agents should keep a record of the `Etag` value returned by the most successful config request, and then present it to future requests via the [If-None-Match](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match) header. If the config has not changed, the server will respond with 304 (Not Modified).
+
+The `Cache-Control` header should contain a `max-age` directive, specifying the amount of time (in seconds) the response should be considered "fresh". Agents should use this to decide how long to wait before requesting config again. The server will respond with a `Cache-Control` header even if the request fails.
+
+#### Dealing with errors
+
+Agents must deal with various error scenarios, including:
+
+ - 7.3 servers where the Kibana connection is not enabled (server responds with 403)
+ - 7.3 servers where the Kibana connection is enabled, but unavailable (server responds with 503)
+ - any other error (server responds with 5xx)
+ - pre-7.3 servers that don't support the config endpoint (server responds with 404)
```

Comment:
oy, another conflicting case is this: 7.3 server where everything is wired up, but there's no data for a service with the given name

Reply:
I don't think this matters that much. In the Node.js agent, we just treat 404 as "try again later", which I think is also what we want in this case.

In the case where the server is 7.3+, the 404 will be accompanied by a `Cache-Control` header and so we'll use its `max-age` value to schedule the retry. In the case of a <7.3 server where no `Cache-Control` header is given, the agent will fall back to a 300 second (5 minutes) retry.

In both cases, no warning or any other info is logged, so this is silent to the user. So I think this is ok.

I think we just need to update this line to say that 404 can happen in both cases.